File: backend/app/routers/graph.py
from datetime import date

from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.dependencies import get_current_user
from app.models.enums import FirmType, TherapeuticArea
from app.schemas.graph import GraphData, TimelineData
from app.services.graph_builder import build_network_graph
from app.services.timeline import build_timeline_data

router = APIRouter()


@router.get("/network", response_model=GraphData)
def get_network_graph(
    firm_type: FirmType | None = None,
    therapeutic_area: TherapeuticArea | None = None,
    from_date: date | None = None,
    to_date: date | None = None,
    lp_committed_only: bool = False,
    db: Session = Depends(get_db),
    _user=Depends(get_current_user),
):
    return build_network_graph(
        db,
        firm_type=firm_type,
        therapeutic_area=therapeutic_area,
        from_date=from_date,
        to_date=to_date,
        lp_committed_only=lp_committed_only,
    )


@router.get("/timeline", response_model=TimelineData)
def get_timeline(
    from_date: date | None = None,
    to_date: date | None = None,
    db: Session = Depends(get_db),
    _user=Depends(get_current_user),
):
    return build_timeline_data(db, from_date=from_date, to_date=to_date)

File: frontend/src/api/graph.ts
import api from './client'
import type { GraphData, TimelineEvent } from '@/types'

export async function getNetworkGraph(params?: {
  firm_type?: string
  therapeutic_area?: string
  from_date?: string
  to_date?: string
  lp_committed_only?: boolean
}): Promise<GraphData> {
  const { data } = await api.get('/graph/network', { params })
  return data
}

export async function getTimeline(params?: {
  from_date?: string
  to_date?: string
}): Promise<{ events: TimelineEvent[] }> {
  const { data } = await api.get('/graph/timeline', { params })
  return data
}

export async function search(q: string) {
  const { data } = await api.get('/search', { params: { q } })
  return data
}
